### ISSUE TITLE
Add `loc=best' kwarg to pyplot.text()

### ISSUE THREAD

**Issue by destroooooy** [matplotlib/matplotlib#1313]
Hi,
  I'm drawing a histogram and I would like to place a text box within the axes that shows the number of events. I like the way I can pass "loc='best'" to pyplot.legend() and it automatically does its best to avoid my data. Can a new kwarg be introduced to pyplot.text() to provide the same functionality?

  ax.text(0.05, 0.95, 'Entries: ' + len(myData),
             transform=ax.transAxes, fontsize=14, verticalalignment='top',
             bbox=dict(boxstyle='square', facecolor='white'))

**Comment by VincentVandalon** [comment]
I would like to give this a go (the generic solution), it will expose me to a new part of the API/code. 

I will read up on it in the next week, investigate if the legend() algorithm can be "borrowed" and how to feed this into text()/annotate() etc. Then, I will post my findings here before starting to implement it.

**Comment by pelson** [comment]
Neat. Good luck @VincentVandalon. Ping here or gitter for any questions, and I'm sure somebody will be able to help. :+1:

**Comment by dashed** [comment]
Is `loc=best' kwarg to pyplot.text() still fine?

**Comment by tacaswell** [comment]
It might be better to build this into `annotation` instead of `text` because `text` takes an `x` and a `y` where as `annotation` already takes an `xy` argument.   Instead of having to add an extra kwarg (which conflicts positional arguments), `xy` can just be allowed to take a string (which is relatively easy to switch on `if is_stringlike(xy) and xy == 'best':` ).

**Comment by dashed** [comment]
@tacaswell Thanks for the tip. Following from your comments, should this issue be closed in favour of https://github.com/matplotlib/matplotlib/issues/4489?

**Comment by nansonzheng** [comment]
Hi @tacaswell, my group is planning to develop this feature with the automatic placement algorithm used in legend.py [(Link to relevant code)](https://github.com/matplotlib/matplotlib/blob/master/lib/matplotlib/legend.py#L921-L980). Is this a good idea to you, or should we come up with a different plan? Do you foresee any problems that may arise?

**Comment by petehuang** [comment]
@tacaswell any guidance on the suggestion to consolidate down to #4489?

**Comment by janosh** [comment]
Is this still part of a milestone? Would certainly be a very useful feature!

**Comment by QuLogic** [comment]
[adjustText](https://github.com/Phlya/adjustText) is a package specifically made to do this sort of thing. I'm not sure we want to go ahead and re-implement it in Matplotlib, but instead if we could make a good reference to it in the documentation, that'd probably be good enough.

**Comment by janosh** [comment]
@QuLogic This seems well within the core functionality of MPL. I don’t think it’s a good idea to expect users to install a separate package just to do this.

**Comment by jklymak** [comment]
@janosh, this would be an interesting project to add to matplotlib - the reason it would be better in the core library is that it could then be incorporated in the draw loop, where it needs to be to properly know the location of every element to avoid.  

But, it would be a pretty big project, and someone would have to care to take it on; probably [adjustText](https://github.com/Phlya/adjustText) is a good start. 

I would be very much against this as a part of `text` - these data-avoiding objects would need to be their own subclass.  

My biggest concern about a project like this is that it would be too complicated, and fragile enough that it would get lots of issues about it, and very few people who understand how it works who are able and willing to maintain it.  Thats why its nice to have bespoke things like this as third-party additions.

**Comment by janosh** [comment]
I see, thanks for taking the time to explain your thoughts! I hope someone comes along who is both able and willing to take this feature on.

**Comment by jklymak** [comment]
Based on the discussion at #12841 I'm going to close this as won't fix.  Coming up with a robust "best" is really hard in general. For a legend we only try 9 locations, so its more clear what "best"  is in that case, and is quite different from "find me some blank space somewhere on the canvas".  

@ImportanceOfBeingErnest had the key insight that the best way forward may be to add this option to AnchoredOffsetbox, which shares legend's 9-position defaults.
